### PR TITLE
Scrape anime and play episodes on browser without ads from terminal

### PR DESCRIPTION
**Pull request by saptarshi1996**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: web_programming/fetch_anime_and_play.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
+from urllib.error import HTTPError
```
Comment (cclauss) [inline web_programming/fetch_anime_and_play.py:1]:
```suggestion
from requests.exceptions import HTTPError, RequestException
```

File: web_programming/fetch_anime_and_play.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+                anime_list.append(
+                    {
+                        "title": anime_title,
+                        "url": anime_url,
+                    }
+                )
+
+        return anime_list
+
+    except (requests.exceptions.RequestException, HTTPError, TypeError) as e:
```
Comment (cclauss) [inline web_programming/fetch_anime_and_play.py:63]:
```suggestion
    except (HTTPError, RequestException, TypeError) as e:
```

File: web_programming/fetch_anime_and_play.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        anime_name (str): [Name of anime]
+
+    Raises:
+        e: [Raises exception on failure]
+
+    Returns:
+        [list]: [List of animes]
+    """
+
+    try:
```
Comment (cclauss) [inline web_programming/fetch_anime_and_play.py:31]: try / except block should be just around critical areas of code.  They should not be 30 lines long!!

File: web_programming/fetch_anime_and_play.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+        episode_url = soup.find("iframe", {"id": "playerframe"})["src"]
+        download_url = episode_url.replace("/embed/", "/playlist/") + ".m3u8"
+        return [f"{BASE_URL}{episode_url}", f"{BASE_URL}{download_url}"]
+
+    except (
+        KeyError,
+        NotFoundErr,
+        TypeError,
+        requests.exceptions.RequestException,
```
Comment (cclauss) [inline web_programming/fetch_anime_and_play.py:157]: Long try / except blocks lead to many different exceptions which confuses the reader of the code.  It also slows down debugging when one of them is raised.

**Review by cclauss — CHANGES_REQUESTED** [review]

File: web_programming/fetch_anime_and_play.py (hunk 1 above)
Comment (saptarshi1996) [reply web_programming/fetch_anime_and_play.py]: changed with from requests.exceptions import RequestException

File: web_programming/fetch_anime_and_play.py (hunk 2 above)
Comment (saptarshi1996) [reply web_programming/fetch_anime_and_play.py]: changed to RequestException

File: web_programming/fetch_anime_and_play.py (hunk 3 above)
Comment (saptarshi1996) [reply web_programming/fetch_anime_and_play.py]: removed the long try catches with critical ones

File: web_programming/fetch_anime_and_play.py (hunk 4 above)
Comment (saptarshi1996) [reply web_programming/fetch_anime_and_play.py]: removed multiple exceptions

**Comment by saptarshi1996** [comment]
@cclauss . I have made the changes

File: web_programming/fetch_anime_and_play.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+    try:
+        response = requests.get(
+            search_url, headers={"UserAgent": UserAgent().chrome}
+        )  # request the url.
+
+        # Is the response ok?
+        response.raise_for_status()
+    except RequestException as e:
+        raise e
```
Comment (cclauss) [inline web_programming/fetch_anime_and_play.py:42]: There is no need to catch an Exception if you are only going to raise the same Exception unmodified.

Comment (saptarshi1996) [reply web_programming/fetch_anime_and_play.py]: ok. i've used response.raise_for_status() instead.

File: web_programming/fetch_anime_and_play.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    episode_list = []
+    for children in episode_page_li:
+        try:
+            if not isinstance(children, NavigableString):
+                episode_list.append(
+                    {
+                        "title": children.find("div", {"class": "name"}).text.replace(
+                            " ", ""
+                        ),
+                        "url": children.find("a")["href"],
```
Comment (cclauss) [inline web_programming/fetch_anime_and_play.py:106]: It confuses the reader that children is plural yet these items are singular.  It is easier to understand when each item is an episode...
```suggestion
    for episode in episode_page_li:
        try:
            if not isinstance(episode, NavigableString):
                episode_list.append(
                    {
                        "title": episode.find("div", {"class": "name"}).text.replace(
                            " ", ""
                        ),
                        "url": episode.find("a")["href"],
```

File: web_programming/fetch_anime_and_play.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+from fake_useragent import UserAgent
+
+BASE_URL = "https://ww1.gogoanime2.org"
+
+
+def search_scraper(anime_name: str) -> list:
+
+    """[summary]
+
+    This function will take an url and
```
Comment (cclauss) [inline web_programming/fetch_anime_and_play.py:14]:
```suggestion
    Take an url and
```

File: web_programming/fetch_anime_and_play.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+                pass
+
+    return anime_list
+
+
+def search_anime_episode_list(episode_endpoint: str) -> list:
+
+    """[summary]
+
+    This function will take an url and
```
Comment (cclauss) [inline web_programming/fetch_anime_and_play.py:69]:
```suggestion
    Take an url and
```

File: web_programming/fetch_anime_and_play.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+                )
+                chosen_episode = episode_list[episode_choice - 1]
+                print(f"You chose {chosen_episode['title']}. Searching...")
+
+                episode_url, download_url = get_anime_episode(chosen_episode["url"])
+                print(f"\nTo watch, ctrl+click on {episode_url}.")
+                print(f"To download, ctrl+click on {download_url}.")
+
+    except (ValueError, IndexError, TypeError) as e:
+        raise e
```
Comment (cclauss) [inline web_programming/fetch_anime_and_play.py:194]: We can lose the try / except if we are not going to handle the errors.

Comment (saptarshi1996) [reply web_programming/fetch_anime_and_play.py]: removed the block.

File: web_programming/fetch_anime_and_play.py (hunk 8 above)
Comment (saptarshi1996) [reply web_programming/fetch_anime_and_play.py]: changed them

File: web_programming/fetch_anime_and_play.py (hunk 6 above)
Comment (saptarshi1996) [reply web_programming/fetch_anime_and_play.py]: for episode in episode_page_li:
for anime in anime_li:
Like this now

File: web_programming/fetch_anime_and_play.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    Returns:
+        [list]: [List of animes]
+    """
+
+    # concat the name to form the search url.
+    search_url = f"{BASE_URL}/search/{anime_name}"
+
+    response = requests.get(
+        search_url, headers={"UserAgent": UserAgent().chrome}
+    )  # request the url.
```
Comment (cclauss) [inline web_programming/fetch_anime_and_play.py:35]: Don't wrap a line just for a comment.
```suggestion
    # request the url.
    response = requests.get(search_url, headers={"UserAgent": UserAgent().chrome})
```

**Review by cclauss — APPROVED** [review]

**Comment by saptarshi1996** [comment]
Thank you @cclauss

**Pull request merged**